### PR TITLE
[XCMP Foreign Fee]: XCMP-Handler accept foreign native token

### PR DESCRIPTION
**Pull request by lreesby**

### CODE REVIEW

File: pallets/automation-time/src/benchmarking.rs [hunk 1]
```diff
@@ -189,8 +189,17 @@ benchmarks! {
 	schedule_xcmp_task_full {
 		let v in 1..T::MaxExecutionTimes::get();
 
+		// TODO: Additional read/writes accounts for AssetRegistry assets unavailable.
```
Comment (lreesby) [inline pallets/automation-time/src/benchmarking.rs:192]: modifying the weight here means no need to manually update github action weights manually.

File: pallets/automation-time/src/lib.rs [hunk 2]
```diff
@@ -1011,7 +1011,15 @@ pub mod pallet {
 				},
 			}
 
+			// TODO: Additional read/writes accounts for AssetRegistry assets unavailable.
```
Comment (lreesby) [inline pallets/automation-time/src/lib.rs:1014]: modifying the weight here means no need to manually update github action weights manually.

File: pallets/automation-time/src/weights.rs [hunk 3]
```diff
@@ -221,13 +221,9 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: ParachainInfo ParachainId (r:1 w:0)
 	// Storage: UnknownTokens ConcreteFungibleBalances (r:1 w:0)
 	// Storage: AssetRegistry LocationToAssetId (r:1 w:0)
```
Comment (lreesby) [inline pallets/automation-time/src/weights.rs:223]: `AssetRegistry` breaks this benchmark but manually adding weights in pallet for now until genesis available.

File: pallets/xcmp-handler/src/lib.rs [hunk 4]
```diff
@@ -434,15 +524,30 @@ pub mod pallet {
...
+		pub fn pay_xcm_fee(
+			currency_id: T::CurrencyId,
+			source: T::AccountId,
+			fee: u128,
+		) -> Result<(), DispatchError> {
 			let local_sovereign_account =
 				Sibling::from(T::SelfParaId::get()).into_account_truncating();
 
-			match T::Currency::transfer(
+			ensure!(
```
Comment (rhuttman) [inline pallets/xcmp-handler/src/lib.rs:535]: Do we really need this? At the moment that code is only called after we have already checked that this action won't kill the account. I know by itself it should use this logic, but we could leave this out and update the ORML repo to have a transfer that does not allow death to eventually be used here

File: pallets/xcmp-handler/src/mock.rs [hunk 5]
```diff
@@ -49,13 +46,16 @@ use xcm_executor::{
...
 pub type Balance = u128;
+pub type BlockNumber = u64;
 pub type CurrencyId = u32;
+pub type TokenId = u32;
 
 pub const ALICE: AccountId32 = AccountId32::new([0u8; 32]);
 pub const LOCAL_PARA_ID: u32 = 2114;
 pub const NATIVE: CurrencyId = 0;
-pub const RELAY: CurrencyId = 1;
+pub const RELAY: CurrencyId = 2;
```
Comment (rhuttman) [inline pallets/xcmp-handler/src/mock.rs:58]: it looks like in `TokenIdConvert` you still expect the relay to be 1

File: pallets/xcmp-handler/src/lib.rs [hunk 6]
```diff
@@ -193,7 +205,7 @@ pub mod pallet {
 		) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
 
-			if currency_id != T::GetNativeCurrencyId::get() {
+			if currency_id != T::GetNativeCurrencyId::get() && xcm_data.native != true {
```
[the remark this answers is not in the document]
Comment (rhuttman) [reply pallets/xcmp-handler/src/lib.rs:208]: No. All Dexes will have more than one native token, and any smart contract platform will have dozens

File: pallets/xcmp-handler/src/lib.rs [hunk 7]
```diff
@@ -354,6 +378,72 @@ pub mod pallet {
...
+			DispatchError,
+		> {
+			let fee_location: MultiLocation = T::CurrencyIdConvert::convert(currency_id)
+				.ok_or(Error::<T>::FailedCurrencyIdToMultiLocation)?;
+
+			let foreign_asset =
+				MultiAsset { id: Concrete(fee_location), fun: Fungibility::Fungible(fee) };
+
+			// This message deducts and burns "amount" from the caller when executed
+			let local_xcm = Xcm(vec![WithdrawAsset::<<T as pallet::Config>::Call>(
```
[the remark this answers is not in the document]
Comment (rhuttman) [reply pallets/xcmp-handler/src/lib.rs:412]: The first answer in this thread has a good explanation for this: https://substrate.stackexchange.com/questions/37/how-can-i-transfer-assets-using-xcm/38

File: pallets/xcmp-handler/src/lib.rs (hunk 4 above)
Comment (lreesby) [reply pallets/xcmp-handler/src/lib.rs:536]: I was doing it out of abundant caution but it is redundant with the current code structure.

**Comment by nvengal** [comment]
Moving to draft for now. Can be reviewed and merged when xcmp foreign fee work is resumed.

**Comment by nvengal** [comment]
outdated

**Pull request closed without merging**